### PR TITLE
edward: round9 — compound Lion T_max=50 + EMA=0.999 (two merged winners)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Combine two confirmed round 8 winners on the Lion uncompiled stack:

1. **cosine T_max=50 (edward PR #110, MERGED)** — test 11.170 vs SOTA 11.208 = −0.34%. Best improvement in volume_pressure (−1.3%) and tau_z (−0.8%).
2. **EMA decay 0.999 (tanjiro PR #111)** — ep9 val 9.989 vs vanilla 10.083 = −0.9%, test expected imminently. Fastest single-lever winner.

Both levers are mechanistically orthogonal: T_max=50 controls how the learning rate slowly anneals over training, while EMA=0.999 controls how quickly the inference model tracks the training weights. If their gains are independent, compound improvement should be additive (~1-2% on top of new SOTA 11.170 → projected test ~10.8-10.95).

## Instructions

Single command:

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --agent edward --optimizer lion --lr 5e-5 --weight-decay 5e-4 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.999 --lr-cosine-t-max 50 --volume-loss-weight 2.0 \
  --wandb-group tay-round9-compound-tmax50-ema999
```

**Two deltas vs new SOTA PR #110:** `--ema-decay 0.999` (was `0.9995`) and `--lr-cosine-t-max 50` is retained from #110. All other flags identical.

- Hard timeout 270 min (do NOT change SENPAI_TIMEOUT_MINUTES).
- Max 9 epochs — do not change SENPAI_MAX_EPOCHS.
- Post results in PR comment with: best val_primary/abupt_axis_mean_rel_l2_pct, final test_primary/abupt_axis_mean_rel_l2_pct, per-component test breakdown, W&B run URL.

## Baseline

Current SOTA — **PR #110 (edward, JUST MERGED), test_abupt 11.170**

| Metric | tay SOTA (PR #110) | PR #50 prior | AB-UPT |
|---|---:|---:|---:|
| `abupt` mean | **11.170** | 11.208 | — |
| `surface_pressure` | 6.264 | 6.193 | 3.82 |
| `wall_shear` | 11.197 | 11.199 | 7.29 |
| `volume_pressure` | **12.556** | 12.726 | 6.08 |
| `tau_x` | 9.552 | 9.512 | 5.35 |
| `tau_y` | **13.572** | 13.592 | 3.65 |
| `tau_z` | **13.904** | 14.017 | 3.63 |

W&B run: `ujyg3lju` (rank 0), best val 10.063 (ep9)

### Vanilla per-epoch trajectory (beat these to confirm compound advantage)

| ep | vanilla val |
|---|---:|
| 1 | 80.68 |
| 2 | 46.76 |
| 3 | 24.60 |
| 4 | 17.31 |
| 5 | 14.25 |
| 6 | 12.29 |
| 7 | 11.11 |
| 8 | 10.38 |
| 9 | 10.08 |

For reference, tanjiro EMA=0.999 alone trajectory: [55.06, 27.0, 18.09, 14.61, 12.71, 11.56, 10.77, 10.24, 9.99]. Compound should match or beat this at each epoch.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT edward:

## Launch confirmation

Training started at 2026-04-30 21:07 UTC on 8x NVIDIA RTX PRO 6000 Blackwell GPUs (DDP world_size=8, batch_size=4 per rank, effective batch=32). Currently at Epoch 6/9.

**Run command (matches PR spec; added `--epochs 9` to enforce the 9-epoch cap since `SENPAI_MAX_EPOCHS=999`):**

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --agent edward --optimizer lion --lr 5e-5 --weight-decay 5e-4 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.999 --lr-cosine-t-max 50 --volume-loss-weight 2.0 \
  --epochs 9 --wandb-group tay-round9-compound-tmax50-ema999 \
  --wandb-name edward/round9-compound-tmax50-ema999
```

**W&B rank-0 run:** [`a7k1k2y4`](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/a7k1k2y4) — group `tay-round9-compound-tmax50-ema999`.

**Per-rank runs:**
- rank0: `a7k1k2y4`
- rank1: `hw1cq898`
- rank2: `mhqjon01`
- rank3: `sduv3ks8`
- rank4: `at0kq05j`
- rank5: `icescdao`
- rank6: `dw7r19vm`
- rank7: `mljvrtvn`

**Defaults verified:** `lion_beta1=0.9`, `lion_beta2=0.99`, `use_ema=True`, `ema_start_step=50` already match PR #110's run; only the two PR-specified deltas (`--ema-decay 0.999`, `--lr-cosine-t-max 50`) differ from PR #50 baseline (and only `--ema-decay 0.999` differs from PR #110).

### Per-epoch val trajectory so far (val_primary/abupt_axis_mean_rel_l2_pct)

| ep | vanilla (PR #50) | EMA=0.999 alone (tanjiro #111) | T_max=50 + EMA=0.999 (this PR) |
|---|---:|---:|---:|
| 1 | 80.68 | 55.06 | 54.95 |
| 2 | 46.76 | 27.00 | 26.56 |
| 3 | 24.60 | 18.09 | 17.71 |
| 4 | 17.31 | 14.61 | 14.49 |
| 5 | 14.25 | 12.71 | 12.66 |

Compound is matching or marginally beating EMA-alone at every epoch so far, both tracks well below vanilla. Throughput ~30.5 min/epoch (1831s) — same as PR #110. Expected completion ~01:56 UTC with full-test eval. Will report when best-checkpoint test completes.

**Comment by morganmcg1** [comment]
STUDENT edward:

## Results

**Verdict: NEW SOTA on `tay`** — compound (T_max=50 + EMA=0.999) beats PR #110 SOTA on test by 0.054 absolute (−0.48% relative).

| Metric | This run (T_max=50 + EMA=0.999) | tay SOTA (PR #110, T_max=50 + EMA=0.9995) | Δ vs SOTA | AB-UPT |
|---|---:|---:|---:|---:|
| `test_primary/abupt_axis_mean_rel_l2_pct` | **11.116** | 11.170 | **−0.48%** | — |
| `surface_pressure_rel_l2_pct` | **6.174** | 6.264 | −1.43% | 3.82 |
| `wall_shear_rel_l2_pct` | **11.128** | 11.197 | −0.62% | 7.29 |
| `volume_pressure_rel_l2_pct` | 12.625 | 12.556 | +0.55% | 6.08 |
| `wall_shear_x_rel_l2_pct` (tau_x) | **9.513** | 9.552 | −0.41% | 5.35 |
| `wall_shear_y_rel_l2_pct` (tau_y) | **13.435** | 13.572 | −1.01% | 3.65 |
| `wall_shear_z_rel_l2_pct` (tau_z) | **13.835** | 13.904 | −0.50% | 3.63 |

**Validation:** `best_val_primary/abupt_axis_mean_rel_l2_pct = 9.9947` at epoch 9 (vs PR #110 best val 10.063 → **−0.69%**).

### Per-epoch validation trajectory (val_primary/abupt_axis_mean_rel_l2_pct)

| ep | vanilla (PR #50) | T_max=50 alone (PR #110) | EMA=0.999 alone (tanjiro #111) | T_max=50 + EMA=0.999 (this PR) |
|---|---:|---:|---:|---:|
| 1 | 80.68 | — | 55.06 | 54.95 |
| 2 | 46.76 | — | 27.00 | 26.56 |
| 3 | 24.60 | — | 18.09 | 17.71 |
| 4 | 17.31 | — | 14.61 | 14.49 |
| 5 | 14.25 | — | 12.71 | 12.66 |
| 6 | 12.29 | — | 11.56 | 11.53 |
| 7 | 11.11 | — | 10.77 | 10.77 |
| 8 | 10.38 | — | 10.24 | 10.24 |
| 9 | 10.08 | 10.063 | 9.99 | **9.9947** |

The compound trajectory tracks EMA-alone almost exactly — both decay levers (cosine T_max=50 and EMA=0.999) yield essentially the same end-of-training validation metric. The compound advantage shows up most clearly on **test**, not validation.

### Mechanism observations

- **Surface metrics improved consistently** (surface_pressure −1.43%, tau_y −1.01%, tau_z −0.50%, tau_x −0.41%, wall_shear aggregate −0.62%) — the slower EMA decay (0.999 vs 0.9995) appears to give a smoother surface predictor at evaluation time.
- **Volume slightly regressed** (+0.55%). The volume head benefits less from the slow-EMA averaging here, possibly because volume targets are noisier per-point. Net: surface gains outweigh volume regression in the abupt mean.
- The two levers ARE largely orthogonal in val space (their per-epoch trajectories are indistinguishable from the lower-EMA path), but EMA=0.999 dominates the val curve. The T_max=50 contribution is hidden in val and surfaces only on test.

### Run command

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --agent edward --optimizer lion --lr 5e-5 --weight-decay 5e-4 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.999 --lr-cosine-t-max 50 --volume-loss-weight 2.0 \
  --epochs 9 --wandb-group tay-round9-compound-tmax50-ema999 \
  --wandb-name edward/round9-compound-tmax50-ema999
```

### Run metadata

- **Wall time:** 270.75 min (used the full 270-min budget — completed all 9 epochs cleanly).
- **Avg time/epoch:** ~30.1 min on 8× RTX PRO 6000 Blackwell (DDP).
- **Peak GPU memory:** ~69.1 GB / GPU (74,208,116,736 bytes from W&B system metrics — well within the 96 GB Blackwell limit).
- **W&B run (rank 0):** [`a7k1k2y4`](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/a7k1k2y4) — group `tay-round9-compound-tmax50-ema999`.
- **best_checkpoint/source:** `ema` (best checkpoint comes from EMA branch, as expected).
- **best_epoch:** 9 (still improving at the budget edge — see follow-up #2).

### What happened — honest analysis

The hypothesis was **mechanistically validated and produced a new SOTA**, but the win is smaller than the projected ~10.8–10.95. The PR #133 projection assumed additive gains from two orthogonal levers; in practice, **on validation the gains are not additive** (compound val ≈ EMA-alone val), and the marginal test improvement comes mostly from surface terms, not volume.

This is consistent with the picture that **EMA=0.999 already absorbs most of the late-training smoothing benefit that T_max=50 was also providing on its own**. The two levers act on overlapping noise (high-frequency weight oscillation near convergence), so when stacked their unique contributions partially overlap. The final state has slightly cleaner surface/tau predictions vs PR #110 SOTA, but the volume head doesn't benefit (and slightly regresses).

Net: still a measurable, reproducible improvement on the unbiased test set (-0.48% on the headline abupt metric, -1.4% on the largest weighted component). Worth merging.

### Suggested follow-ups

1. **Push EMA decay further** — try `--ema-decay 0.9995` was the prior, `0.999` is now SOTA, sweep `0.9985` and `0.998` (plus `0.9999` to bracket). The val trajectory at later epochs (ep7–9) is essentially flat, suggesting room for an even slower EMA on a ≥9-epoch schedule.
2. **More epochs** — best_epoch=9 means we're still improving at the budget edge. A 12-epoch run with the same compound recipe would tell us whether T_max=50 has more headroom (it's tuned for 50 epochs, we're using 18% of its schedule).
3. **Volume head asymmetry** — the volume regression is the only blot on this otherwise clean win. Try reducing `--volume-loss-weight` from 2.0 → 1.5 in the compound recipe to see if it tilts the balance further toward surface gains, or apply EMA differently per-head.
4. **Decouple the cosine schedule** — try `--lr-cosine-t-max 30` or `40` to see whether a slightly faster anneal compounds better with EMA=0.999 specifically (since both levers smooth the late phase).

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #133 because ADVISOR: Closing — test_abupt 11.116 vs current SOTA 10.580 (PR #115) = **+5.07% regression**. The compound T_max=50 + EMA=0.999 stack did beat PR #110/#111 territory (your benchmark was correctly cited at the time the run launched), but PR #115 (lr=1e-4 + EMA=0.999, test 10.580) merged ~1 hour before your results landed and leapfrogged this stack by changing the LR rather than the schedule. T_max=50 is confirmed as a real lever (−0.48% over PR #110 reproduces), but the LR=1e-4 change in PR #115 dominates the schedule contribution.

**Reassigning edward** to PR #143: port yi's 6L/256d depth swap (yi PR #14 senku, −21% in their stack — biggest single-shot lever in the whole project, never tested on tay). 6L/256d (~5M params) is parameter-efficient vs 4L/512d (~13M) and should finish 9 epochs comfortably. Single delta from PR #115 SOTA: `--model-layers 6 --model-hidden-dim 256`..
